feat: Promote send-via-https as generic utility

splasky opened the pull request:
Defining `send_http_msg()` inside `main.c` may cause us hard to intergrate with legato or implement unit-test. A new file `url.h` is added, it includes HTTPS protocol implementation.

Change `send_http_msg()` prototype for sending general message instead of restricting format.
    
Closes #49

howjmay reviewed (changes requested):
Change git message to "Moving `send_http_msg` under `/utils` makes generating library and being used by other devices easier."

howjmay commented on utils/Makefile:
```diff
@@ -1,15 +1,9 @@
+	$(CC) -c $(CFLAGS) $(INCLUDES) -o $@ $^
 
 clean:
 	rm -f *.o
```
Add a newline here

howjmay commented on utils/url.c:
```diff
@@ -0,0 +1,35 @@
+  http_parser_init(parser, HTTP_RESPONSE);
+  http_parser_execute(parser, &settings, res, strlen(res));
+#ifdef DEBUG
+  printf("HTTP Response: %s\n", http_res_body);
```
？Where is the variable `http_res_body` declared？

splasky replied:
Global variable which is defined inside `conn_http.h`.

howjmay commented on utils/url.h:
```diff
@@ -0,0 +1,27 @@
+
+#include "defined_error.h"
+/**
+ * @brief send message vi http protocal
```
via

howjmay commented on utils/url.h:
```diff
@@ -0,0 +1,27 @@
+/**
+ * @brief send message vi http protocal
+ *
+ * @param host http host
```
@param[in] host http host

howjmay reviewed (changes requested):
we use rebase instead of merge

Commit:
feat: Promote send-via-https as generic utility

Defining send_http_msg() inside main.c may cause us hard to intergrate
with legato or implement unit-test. A new file url.h is added,
it includes https protocol implementation.

Change send_http_msg() prototype for sending general message instead of restricting format.

Closes #49

jserv commented on Makefile:
```diff
@@ -29,8 +29,8 @@ INCLUDES = -I$(THIRD_PARTY_PATH)/http-parser -I$(MBEDTLS_PATH)/include -I$(ROOT_
 
-UTILS_OBJS = $(UTILS_PATH)/crypto_utils.o $(UTILS_PATH)/serializer.o $(UTILS_PATH)/tryte_byte_conv.o $(UTILS_PATH)/uart_utils.o
-# We need to modify this rule here to be compatible to the situation 
+UTILS_OBJS = $(UTILS_PATH)/crypto_utils.o $(UTILS_PATH)/serializer.o $(UTILS_PATH)/tryte_byte_conv.o $(UTILS_PATH)/uart_utils.o $(UTILS_PATH)/url.o
```
Split into two line for pretty print.

jserv commented:
Since the function prototype of `send_http_msg` was changed, the subject is no longer accurate. Rework!

jserv commented:
Don't abuse "Refactor." Please do look up dictionary and improve the subject. You can use more words though.

jserv commented:
Check [Google's Engineering Practices documentation](https://google.github.io/eng-practices/review/) for understanding fundamental concepts.

howjmay reviewed (changes requested):
typo `protype` should be `prototype`

howjmay commented on utils/url.c:
```diff
@@ -0,0 +1,35 @@
+  set_post_request(api, host, atoi(port), msg, &req);
+  http_parser_settings settings;
+  settings.on_body = parser_body_callback;
+  http_parser *parser = malloc(sizeof(http_parser));
```
Why don't you statically declare the variable, `parser`?

howjmay commented on utils/url.h:
```diff
@@ -0,0 +1,27 @@
+
+#include "defined_error.h"
+/**
+ * @brief send message via http protocal
```
protocol

howjmay commented on utils/url.h:
```diff
@@ -0,0 +1,27 @@
+ * @param[in] ssl_seed seed for ssl connection
+ *
+ * @return http status
+ * @see http_retcode_t
```
retcode_t

jserv commented:
Mention the issue number that pull request is going to close.

jserv commented on utils/url.c:
```diff
@@ -0,0 +1,35 @@
+#include "url.h"
```
Add copyright notice.

jserv commented:
It does not make sense for the following message
> Move send_http_msg() under utils for reuse for other devices.

Reasons:
1. Lack of the descriptions about example devices;
2. Fix grammar;
3. No information on `send_http_msg`.

jserv commented:
Fix the grammar again in PR.

howjmay reviewed (changes requested):
Defining `send_http_msg()` inside `main.c` may cause us hard to integrate with legato or implement unit-test. A new file `url.h` is added, it includes https protocol implementation.

Change `send_http_msg()` prototype for sending general message instead of
restricting format.

howjmay commented on utils/url.h:
```diff
@@ -0,0 +1,29 @@
+ * @param[in] msg_len length of message
+ * @param[in] ssl_seed seed for ssl connection
+ *
+ * @return http status
```
Follow the document format we did in tangle-accelerator in the `@return` field.

jserv commented:
The subject is misleading. Indeed, it is a matter of https-oriented sender routine.

howjmay commented on utils/url.c:
```diff
@@ -0,0 +1,43 @@
+                         const char *ssl_seed) {
+  char res[4096] = {0};
+  char *req = NULL;
+  retcode_t ret = 0;
```
Initialize `ret` with enum but not integer.

jserv commented:
Avoid using "Modify" in the subject unless you just perform unwanted changes.

jserv reviewed (changes requested):
Change the subject for the sake of revised one.

jserv commented on utils/url.h:
```diff
@@ -0,0 +1,28 @@
+ *
+ * @param[in] host http host
+ * @param[in] port http port
+ * @param[in] api http api
```
The descriptions about `api` is misleading. You have to explain with informative wording.

jserv commented on utils/url.h:
```diff
@@ -0,0 +1,28 @@
+ *
+ * @param[in] host https host
+ * @param[in] port https port
+ * @param[in] api https api for post request
```
The type of `api` is `char *`, which might not reflect what it behave as "https api." You shall clarify.

jserv replied:
Also, according to the usage, this file should be renamed to `utils/protocol.h` rather than `utils/url.h`.

splasky replied:
Is it better to concatenate host and api?

jserv replied:
It is indeed referring to host, rather than generic api.
